### PR TITLE
Include unpartitioned elements in MeshBase::subdomain_ids

### PR DESCRIPTION
**Pull request by loganharbour**
Resolves https://github.com/idaholab/moose/issues/17245.

### CODE REVIEW

File: tests/mesh/mesh_generation_test.C [hunk 1]
```diff
@@ -86,10 +86,17 @@ public:
...
+
     CPPUNIT_ASSERT_EQUAL(mesh.n_elem(), cast_int<dof_id_type>(n));
     CPPUNIT_ASSERT_EQUAL(mesh.n_nodes(),
                          cast_int<dof_id_type>((Elem::type_to_n_nodes_map[type]-1)*n + 1));
 
+   // Make sure that subdomain ids for unpartitioned elements are included
+   std::set<subdomain_id_type> ids;
+   mesh.subdomain_ids(ids);
+   CPPUNIT_ASSERT_EQUAL(ids.size(), (std::size_t)1);
+   CPPUNIT_ASSERT_EQUAL(*ids.begin(), (subdomain_id_type)0;
```
Comment (jwpeterson) [inline tests/mesh/mesh_generation_test.C:98]: This line (and a similar one below) is missing a trailing parenthesis?

Comment (loganharbour) [reply tests/mesh/mesh_generation_test.C]: Yeah - I pushed those in haste before testing and actually removed. 

I thought I could get away with testing this with the MeshTools::Generation meshes, but I forgot that we call prepare_for_use() at the end of all of those calls. Therefore, I wouldn't really be testing as we'd partition.

**Comment by loganharbour** [comment]
I could possibly add a unit test that sets `skip_partitioning(true)` and uses `MeshTools::Generation` to build a simple mesh. Should be good then to make sure subdomain 0 exits in the mesh via the (all) unpartitioned elements.

Thoughts? Alternative?

**Comment by roystgnr** [comment]
Code looks good.  Would it make sense to just create a mesh by hand in a unit test?  We've got a few like that in e.g. the slit mesh and the multi-dimensional mesh unit tests.

File: tests/mesh/mesh_subdomain_id.C [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  CPPUNIT_TEST_SUITE_END();
+
+public:
+  void setUp() {}
+
+  void tearDown() {}
+
+  void testUnpartitioned()
+  {
+    std::unique_ptr<UnstructuredMesh> mesh = libmesh_make_unique<ReplicatedMesh>(*TestCommWorld);
```
Comment (roystgnr) [inline tests/mesh/mesh_subdomain_id.C:33]: Why force ReplicatedMesh here?  If we use Mesh it'll test replicated or distributed depending on configure settings, and we'd expect it to work either way.

**Review by roystgnr — APPROVED** [review]

**Comment by loganharbour** [comment]
Should be good to merge.

**Pull request merged**